imxrt-hal/src/can/mod.rs

```diff
@@ -0,0 +1,972 @@
+        M::USIZE == 2
+    }
+
+    pub(crate) fn base_address(&self) -> u32 {
+        let addr: *const ral::can::RegisterBlock = &*self.reg;
+        addr as u32
+    }
+
+    pub fn free(self) -> ral::can::Instance {
+        self.reg
+    }
+
+    fn soft_reset(&mut self) {
+        ral::modify_reg!(ral::can, self.reg, MCR, SOFTRST: SOFTRST_1);
+        while ral::read_reg!(ral::can, self.reg, MCR, SOFTRST == SOFTRST_1) {}
+    }
+
+    fn enter_freeze_mode(&mut self) {
+        ral::modify_reg!(ral::can, self.reg, MCR, FRZ: FRZ_1);
+        ral::modify_reg!(ral::can, self.reg, MCR, HALT: HALT_1);
+        while ral::read_reg!(ral::can, self.reg, MCR, FRZACK != FRZACK_1) {}
+    }
+
+    fn exit_freeze_mode(&mut self) {
+        ral::modify_reg!(ral::can, self.reg, MCR, HALT: HALT_0);
+        while ral::read_reg!(ral::can, self.reg, MCR, FRZACK != FRZACK_0) {}
+    }
+
+    pub fn set_mrp(&mut self, mrp: bool) {
+        ral::modify_reg!(ral::can, self.reg, CTRL2, MRP: mrp as u32)
+    }
+
+    pub fn set_rrs(&mut self, rrs: bool) {
+        ral::modify_reg!(ral::can, self.reg, CTRL2, RRS: rrs as u32)
+    }
+
+    pub fn get_clock(&self) -> u32 {
+        self.source_clock.0
+    }
+
+    fn result_to_bit_table(&self, result: u8) -> Option<[u32; 3]> {
+        match result {
+            0 => Some([0, 0, 1]),
+            1 => Some([1, 0, 1]),
+            2 => Some([1, 1, 1]),
+            3 => Some([2, 1, 1]),
+            4 => Some([2, 2, 1]),
+            5 => Some([2, 3, 1]),
+            6 => Some([2, 3, 2]),
+            7 => Some([2, 4, 2]),
+            8 => Some([2, 5, 2]),
+            9 => Some([2, 5, 3]),
+            10 => Some([2, 6, 3]),
+            11 => Some([2, 7, 3]),
+            12 => Some([2, 7, 4]),
+            13 => Some([3, 7, 4]),
+            14 => Some([3, 7, 5]),
+            15 => Some([4, 7, 5]),
+            16 => Some([4, 7, 6]),
+            17 => Some([5, 7, 6]),
+            18 => Some([6, 7, 6]),
+            19 => Some([6, 7, 7]),
+            20 => Some([7, 7, 7]),
+            _ => None,
+        }
+    }
+
+    pub fn set_baud_rate(&mut self, baud: u32) {
+        let clock_freq = self.get_clock();
+
+        let mut divisor = 0;
+        let mut best_divisor: u32 = 0;
+        let mut result: u32 = clock_freq / baud / (divisor + 1);
+        let mut error: i16 = (baud - (clock_freq / (result * (divisor + 1)))) as i16;
+        let mut best_error = error;
+
+        self.while_frozen(|this| {
+            while result > 5 {
+                divisor += 1;
+                result = clock_freq / baud / (divisor + 1);
+                if result <= 25 {
+                    error = (baud - (clock_freq / (result * (divisor + 1)))) as i16;
+                    if error < 0 {
+                        error *= -1;
+                    }
+                    if error < best_error {
+                        best_error = error;
+                        best_divisor = divisor;
+                    }
```

Comment:
Unsigned subtraction of `baud` and `clock_freq / [...]` -- before casting to `i16` -- can overflow. This can cause a panic when building with default _debug_ configurations.

```
[ERROR teensy4_panic]: panicked at 'attempt to subtract with overflow', /path/to/imxrt-hal/src/can/mod.rs:370:29
```

I can reproduce the panic in the teensy4-rs `can.rs` example (as of [this commit](https://github.com/mciantyre/teensy4-rs/pull/124/commits/d176a6d17ea2c7e022aad9d21461168d3cc59473)). The snippet below shows how I'm building the example. I haven't stepped through the code to understand what specific iteration produces this panic.

```
cargo build --target=thumbv7em-none-eabihf --example=can --all-features
```

My first thought is to use something like [`abs_diff`](https://doc.rust-lang.org/std/primitive.u32.html#method.abs_diff). However, if the absolute difference is larger than what fits in an `i16`, we'll truncate that difference once we cast from `u32` to `i16`.
- Could we revise this routine so that it's less likely to panic in debug builds?
- Are the truncating casts to `i16` necessary?

Reply:
This has been updated to use the `abs_diff` method.